File: code/scenes/src/main/java/org/fbme/scenes/controllers/SceneViewpointByCell.java
package org.fbme.scenes.controllers;

import jetbrains.mps.openapi.editor.cells.EditorCell;
import org.fbme.scenes.controllers.scene.SceneEditor;
import org.fbme.scenes.controllers.scene.SceneInitializer;
import org.fbme.scenes.controllers.scene.ViewpointAndScaleData;

import java.awt.*;

public class SceneViewpointByCell implements ModifiableSceneViewpoint {

    private final EditorCell myCell;
    private final SceneEditor myEditor;

    private int shiftX;
    private int shiftY;

    public SceneViewpointByCell(SceneEditor editor, EditorCell cell) {
        myCell = cell;
        myEditor = editor;

        editor.addInitializer(new MyInitializer());
    }

    @Override
    public Point getEditorShift() {
        return new Point(myCell.getX() + shiftX, myCell.getY() + shiftY);
    }

    @Override
    public void setEditorShift(Point origin) {
        shiftX = origin.x - myCell.getX();
        shiftY = origin.y - myCell.getY();
        myEditor.fireRelayout();
    }

    @Override
    public float getEditorScale() {
        int lineSize = LayoutUtil.getLineSize(myCell.getStyle());
        return scale(lineSize);
    }

    private float scale(int lineSize) {
        return ((float) lineSize) / 100;
    }

    public class MyInitializer implements SceneInitializer {

        @Override
        public void onAdd() {
            ViewpointAndScaleData data = myEditor.loadState(ViewpointAndScaleData.KEY);
            if (data != null) {
                shiftX = data.shiftX;
                shiftY = data.shiftY;
                LayoutUtil.setFontSize(myCell.getStyle(), data.lineSize);
            }
        }

        @Override
        public void onRemove() {
            myEditor.storeState(ViewpointAndScaleData.KEY, new ViewpointAndScaleData(shiftX, shiftY, LayoutUtil.getFontSize(myCell.getStyle())));
        }
    }

    public EditorCell getCell() {
        return myCell;
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/SceneViewpointByCell.kt
package org.fbme.scenes.controllers

import jetbrains.mps.openapi.editor.cells.EditorCell
import org.fbme.scenes.controllers.scene.SceneEditor
import org.fbme.scenes.controllers.scene.SceneInitializer
import org.fbme.scenes.controllers.scene.ViewpointAndScaleData
import java.awt.Point

class SceneViewpointByCell(
    private val myEditor: SceneEditor,
    val cell: EditorCell,
    private var shiftX: Int = 0,
    private var shiftY: Int = 0
) : SceneViewpoint {
    override var editorShift: Point = Point(cell.x + shiftX, cell.y + shiftY)
        get() {
            return Point(cell.x + shiftX, cell.y + shiftY)
        }
        set(value) {
            field.x = value.x - cell.x
            field.y = value.y - cell.y
            myEditor.fireRelayout()
        }

    override val editorScale: Float
        get() {
            return LayoutUtil.getScale(cell.style)
        }

    inner class MyInitializer : SceneInitializer {
        override fun onAdd() {
            val data = myEditor.loadState(ViewpointAndScaleData.KEY)
            if (data != null) {
                shiftX = data.shiftX
                shiftY = data.shiftY
                LayoutUtil.setFontSize(cell.style, data.fontSize)
            }
        }

        override fun onRemove() {
            myEditor.storeState(
                ViewpointAndScaleData.KEY,
                ViewpointAndScaleData(shiftX, shiftY, LayoutUtil.getFontSize(cell.style))
            )
        }
    }

    init {
        myEditor.addInitializer(MyInitializer())
    }
}
